File: app/src/main/java/com/veganbeauty/app/features/community/com_feed/CommunityExploreFragment.java
package com.veganbeauty.app.features.community.com_feed;



import android.os.Bundle;

import android.os.Handler;

import android.os.Looper;

import android.view.LayoutInflater;

import android.view.View;

import android.view.ViewGroup;

import androidx.annotation.NonNull;

import androidx.annotation.Nullable;

import androidx.lifecycle.ViewModelProvider;

import androidx.recyclerview.widget.RecyclerView;

import androidx.viewpager2.widget.ViewPager2;

import com.veganbeauty.app.R;

import com.veganbeauty.app.core.base.RootieFragment;

import com.veganbeauty.app.data.local.LocalJsonReader;

import com.veganbeauty.app.data.local.RootieDatabase;

import com.veganbeauty.app.data.remote.FirestoreService;

import com.veganbeauty.app.data.repository.CommunityRepository;

import com.veganbeauty.app.databinding.ComFragmentExploreBinding;

import com.veganbeauty.app.data.local.entities.YtVideoEntity;

import com.veganbeauty.app.utils.ComBottomNavHelper;

import com.veganbeauty.app.utils.ExploreVideoCache;

import java.util.ArrayList;

import java.util.Collections;

import java.util.List;



public class CommunityExploreFragment extends RootieFragment {



    private ComFragmentExploreBinding _binding;



    private CommunityViewModel viewModel;

    private ExploreVideoAdapter exploreAdapter = new ExploreVideoAdapter(new ArrayList<>(), null);



    private boolean isNavVisible = true;

    private ExploreVideoAdapter.VideoViewHolder currentPlayingHolder;

    private List<YtVideoEntity> currentVideos = Collections.emptyList();



    private void hideBottomNavigation() {

        if (!isNavVisible) return;

        isNavVisible = false;

        float height = _binding.comBottomNav.getRoot().getHeight();

        _binding.comBottomNav.getRoot().animate()

            .translationY(height)

            .setInterpolator(new android.view.animation.AccelerateDecelerateInterpolator())

            .setDuration(250)

            .start();



        exploreAdapter.setNavBarVisible(false, true);

    }



    private void showBottomNavigation() {

        if (isNavVisible) return;

        isNavVisible = true;

        _binding.comBottomNav.getRoot().animate()

            .translationY(0f)

            .setInterpolator(new android.view.animation.AccelerateDecelerateInterpolator())

            .setDuration(250)

            .start();



        exploreAdapter.setNavBarVisible(true, true);

    }



    private void playVideoAtPosition(int position) {

        exploreAdapter.setCurrentPosition(position);

        if (currentPlayingHolder != null) {

            currentPlayingHolder.pauseVideo();

            currentPlayingHolder = null;

        }



        RecyclerView rv = (RecyclerView) _binding.viewPagerExplore.getChildAt(0);

        if (rv == null) return;



        RecyclerView.ViewHolder holder = rv.findViewHolderForAdapterPosition(position);

        if (holder instanceof ExploreVideoAdapter.VideoViewHolder) {

            ExploreVideoAdapter.VideoViewHolder videoHolder = (ExploreVideoAdapter.VideoViewHolder) holder;

            videoHolder.playVideo();

            currentPlayingHolder = videoHolder;

        } else {

            new Handler(Looper.getMainLooper()).postDelayed(() -> playVideoAtPosition(position), 150);

        }

    }



    private void prefetchAround(int position) {

        if (currentVideos.isEmpty()) return;

        if (!ExploreVideoCache.isNetworkAvailable(requireContext())) return;



        for (int offset = 0; offset <= 2; offset++) {

            int index = position + offset;

            if (index >= 0 && index < currentVideos.size()) {

                ExploreVideoCache.prefetch(requireContext(), currentVideos.get(index).getUrl());

            }

        }

    }



    @Nullable

    @Override

    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        _binding = ComFragmentExploreBinding.inflate(inflater, container, false);

        setupViewModel();

        return _binding.getRoot();

    }



    private void setupViewModel() {

        RootieDatabase db = RootieDatabase.getDatabase(requireContext());

        CommunityRepository repository = new CommunityRepository(db.communityDao(), new LocalJsonReader(requireContext()), new FirestoreService());

        CommunityViewModelFactory factory = new CommunityViewModelFactory(repository);

        viewModel = new ViewModelProvider(requireActivity(), factory).get(CommunityViewModel.class);

    }



    @Override

    protected void setupUI(@NonNull View view) {

        int navMargin = getResources().getDimensionPixelSize(R.dimen.com_content_padding_bottom);

        int compactMargin = getResources().getDimensionPixelSize(R.dimen.com_explore_content_margin_bottom_compact);

        exploreAdapter.setBottomMargins(navMargin, compactMargin);



        _binding.viewPagerExplore.setAdapter(exploreAdapter);

        _binding.viewPagerExplore.setOffscreenPageLimit(1);



        _binding.viewPagerExplore.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {

            private int lastPosition = 0;



            @Override

            public void onPageSelected(int position) {

                super.onPageSelected(position);



                if (position > lastPosition) {

                    hideBottomNavigation();

                } else if (position < lastPosition) {

                    showBottomNavigation();

                }

                lastPosition = position;



                playVideoAtPosition(position);

                prefetchAround(position);

            }

        });



        _binding.ivBack.setOnClickListener(v -> ComBottomNavHelper.navigateToAppHome(this));



        _binding.ivSearch.setOnClickListener(v -> {

            getParentFragmentManager().beginTransaction()

                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)

                .replace(R.id.main_container, new ExploreSearchFragment())

                .addToBackStack(null)

                .commit();

        });



        ComBottomNavHelper.setup(this, _binding.comBottomNav.getRoot(), ComBottomNavHelper.TAB_EXPLORE);

    }



    @Override

    protected void observeViewModel() {

        viewModel.getExploreVideos().observe(getViewLifecycleOwner(), videos -> {

            List<YtVideoEntity> safeVideos = videos != null ? videos : Collections.emptyList();

            List<YtVideoEntity> finalVideos = new ArrayList<>(safeVideos);

            Collections.shuffle(finalVideos);



            String targetId = getArguments() != null ? getArguments().getString("target_video_id") : null;

            if (targetId != null) {

                YtVideoEntity targetVideo = null;

                for (YtVideoEntity v : finalVideos) {

                    if (v.getId().equals(targetId)) {

                        targetVideo = v;

                        break;

                    }

                }

                if (targetVideo != null) {

                    finalVideos.remove(targetVideo);

                    finalVideos.add(0, targetVideo);

                }

            }



            currentVideos = finalVideos;

            exploreAdapter.updateData(finalVideos);



            if (!finalVideos.isEmpty()) {

                new Handler(Looper.getMainLooper()).postDelayed(() -> {

                    playVideoAtPosition(0);

                    prefetchAround(0);

                }, 300);

            }

        });

    }



    @Override

    public void onDestroyView() {

        currentPlayingHolder = null;

        super.onDestroyView();

        _binding = null;

    }

}

File: app/src/main/java/com/veganbeauty/app/features/community/com_feed/CommunityExploreFragment.kt
package com.veganbeauty.app.features.community.com_feed

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import androidx.room.Room
import androidx.viewpager2.widget.ViewPager2
import com.veganbeauty.app.R
import com.veganbeauty.app.core.base.RootieFragment
import com.veganbeauty.app.data.local.LocalJsonReader
import com.veganbeauty.app.data.local.RootieDatabase
import com.veganbeauty.app.data.remote.FirestoreService
import com.veganbeauty.app.data.repository.CommunityRepository
import com.veganbeauty.app.databinding.ComFragmentExploreBinding
import com.veganbeauty.app.features.home.HomeFragment

class CommunityExploreFragment : RootieFragment() {

    private var _binding: ComFragmentExploreBinding? = null
    private val binding get() = _binding!!

    private lateinit var viewModel: CommunityViewModel
    private val exploreAdapter = ExploreVideoAdapter(emptyList())

    private var isNavVisible = true

    private fun hideBottomNavigation() {
        if (!isNavVisible) return
        isNavVisible = false
        val height = binding.comBottomNav.root.height.toFloat()
        binding.comBottomNav.root.animate()
            .translationY(height)
            .setInterpolator(android.view.animation.AccelerateDecelerateInterpolator())
            .setDuration(250)
            .start()
        
        updateVideoContentTranslation(height)
    }

    private fun showBottomNavigation() {
        if (isNavVisible) return
        isNavVisible = true
        binding.comBottomNav.root.animate()
            .translationY(0f)
            .setInterpolator(android.view.animation.AccelerateDecelerateInterpolator())
            .setDuration(250)
            .start()
            
        updateVideoContentTranslation(0f)
    }

    private fun updateVideoContentTranslation(translationY: Float) {
        exploreAdapter.contentTranslationY = translationY
        val rv = binding.viewPagerExplore.getChildAt(0) as? RecyclerView
        val layoutManager = rv?.layoutManager as? androidx.recyclerview.widget.LinearLayoutManager
        layoutManager?.let {
            val first = it.findFirstVisibleItemPosition()
            val last = it.findLastVisibleItemPosition()
            if (first != -1 && last != -1) {
                for (i in first..last) {
                    val holder = rv.findViewHolderForAdapterPosition(i) as? ExploreVideoAdapter.ExploreVideoViewHolder
                    holder?.animateContent(translationY)
                }
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = ComFragmentExploreBinding.inflate(inflater, container, false)
        setupViewModel()
        return binding.root
    }

    private fun setupViewModel() {
        val db = RootieDatabase.getDatabase(requireContext())
        val repository = CommunityRepository(db.communityDao(), LocalJsonReader(requireContext()), FirestoreService())
        val factory = CommunityViewModelFactory(repository)
        viewModel = ViewModelProvider(this, factory)[CommunityViewModel::class.java]
    }

    override fun setupUI(view: View) {
        // Setup ViewPager2
        binding.viewPagerExplore.adapter = exploreAdapter
        binding.viewPagerExplore.offscreenPageLimit = 1

        binding.viewPagerExplore.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            private var currentPlayingHolder: ExploreVideoAdapter.ExploreVideoViewHolder? = null
            private var lastPosition = 0

            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                
                // Handle nav bar visibility based on scroll direction
                if (position > lastPosition) {
                    hideBottomNavigation()
                } else if (position < lastPosition) {
                    showBottomNavigation()
                }
                lastPosition = position
                
                // Stop previous
                currentPlayingHolder?.stopVideo()
                
                // Play new
                val rv = binding.viewPagerExplore.getChildAt(0) as? RecyclerView
                val holder = rv?.findViewHolderForAdapterPosition(position) as? ExploreVideoAdapter.ExploreVideoViewHolder
                if (holder != null) {
                    holder.playVideo()
                    currentPlayingHolder = holder
                } else {
                    // If holder is not ready yet, delay a bit
                    Handler(Looper.getMainLooper()).postDelayed({
                        val delayedHolder = rv?.findViewHolderForAdapterPosition(position) as? ExploreVideoAdapter.ExploreVideoViewHolder
                        delayedHolder?.playVideo()
                        currentPlayingHolder = delayedHolder
                    }, 200)
                }
            }
        })

        // Setup UI Actions
        binding.ivBack.setOnClickListener {
            parentFragmentManager.beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.main_container, CommunityFeedFragment())
                .commit()
        }

        binding.ivSearch.setOnClickListener {
            parentFragmentManager.beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.main_container, ExploreSearchFragment())
                .addToBackStack(null)
                .commit()
        }

        // Setup Bottom Nav Navigation
        binding.comBottomNav.navComFeed.setOnClickListener {
            parentFragmentManager.beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.main_container, CommunityFeedFragment())
                .commit()
        }

        binding.comBottomNav.navComProfile.setOnClickListener {
            parentFragmentManager.beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.main_container, com.veganbeauty.app.features.community.profile.CommunityProfileFragment())
                .commit()
        }

        binding.comBottomNav.navComHub.setOnClickListener {
            parentFragmentManager.beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.main_container, com.veganbeauty.app.features.community.beauty_hub.CommunityBeautyHubFragment())
                .commit()
        }

        binding.comBottomNav.navComChat.setOnClickListener {
            parentFragmentManager.beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.main_container, com.veganbeauty.app.features.community.message.CommunityMessageFragment())
                .commit()
        }

        // Keep Khám phá active visually
        val exploreIcon = binding.comBottomNav.navComExplore.getChildAt(0) as? android.widget.ImageView
        exploreIcon?.setColorFilter(resources.getColor(R.color.primary, null))
        val exploreText = binding.comBottomNav.navComExplore.getChildAt(1) as? android.widget.TextView
        exploreText?.setTextColor(resources.getColor(R.color.primary, null))
        exploreText?.setTypeface(null, android.graphics.Typeface.BOLD)

        val feedIcon = binding.comBottomNav.navComFeed.getChildAt(0) as? android.widget.ImageView
        feedIcon?.setColorFilter(resources.getColor(R.color.tertiary, null))
        val feedText = binding.comBottomNav.navComFeed.getChildAt(1) as? android.widget.TextView
        feedText?.setTextColor(resources.getColor(R.color.tertiary, null))
        feedText?.setTypeface(null, android.graphics.Typeface.NORMAL)
    }

    override fun observeViewModel() {
        viewModel.exploreVideos.observe(viewLifecycleOwner) { videos ->
            var finalVideos = videos.shuffled()
            
            val targetId = arguments?.getString("target_video_id")
            if (targetId != null) {
                val targetVideo = finalVideos.find { it.id == targetId }
                if (targetVideo != null) {
                    val listWithoutTarget = finalVideos.filter { it.id != targetId }.toMutableList()
                    listWithoutTarget.add(0, targetVideo)
                    finalVideos = listWithoutTarget
                }
            }

            exploreAdapter.updateData(finalVideos)
            
            // Start playing first video if possible
            if (finalVideos.isNotEmpty()) {
                Handler(Looper.getMainLooper()).postDelayed({
                    val rv = binding.viewPagerExplore.getChildAt(0) as? RecyclerView
                    val holder = rv?.findViewHolderForAdapterPosition(0) as? ExploreVideoAdapter.ExploreVideoViewHolder
                    holder?.playVideo()
                }, 500)
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}

